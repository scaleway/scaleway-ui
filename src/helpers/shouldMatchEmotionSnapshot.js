/* eslint-disable import/no-extraneous-dependencies */
import createCache from '@emotion/cache'
import { createSerializer } from '@emotion/jest'
import { CacheProvider, ThemeProvider } from '@emotion/react'
import { render } from '@testing-library/react'
import React from 'react'
import theme from '../theme'

expect.addSnapshotSerializer(createSerializer())

const emotionCache = createCache({
  key: 'cache',
})
emotionCache.compat = true

export default async (component, { transform } = {}) => {
  const node = render(
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={theme}>{component}</ThemeProvider>
    </CacheProvider>,
  )
  if (transform) await transform(node)

  expect(node.asFragment()).toMatchSnapshot()
}
