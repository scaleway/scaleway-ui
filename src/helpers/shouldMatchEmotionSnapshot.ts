/* eslint-disable import/no-extraneous-dependencies */
import { createSerializer } from '@emotion/jest'
import { RenderOptions, render } from '@testing-library/react'
import { ReactNode } from 'react'
import defaultTheme from '../theme'
import renderWithTheme from './renderWithTheme'

// use only class hash (generated from css style content)
expect.addSnapshotSerializer(
  createSerializer({ classNameReplacer: className => className }),
)

interface Options {
  options?: RenderOptions
  theme?: typeof defaultTheme
  transform?: (node: ReturnType<typeof render>) => Promise<void> | void
}

export default async (
  component: ReactNode,
  { transform, options, theme }: Options = {},
): Promise<void> => {
  const node = renderWithTheme(component, options, theme)
  if (transform) await transform(node)

  expect(node.asFragment()).toMatchSnapshot()
  node.unmount()
}
