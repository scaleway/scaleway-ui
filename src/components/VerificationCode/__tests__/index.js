import userEvent from '@testing-library/user-event'
import React from 'react'
import { VerificationCode } from '..'
import shouldMatchEmotionSnapshot from '../../../helpers/shouldMatchEmotionSnapshot'

const pasteEventWithValue = (selector, value) => {
  userEvent.paste(selector, value, {
    clipboardData: {
      getData: () => value,
    },
  })
}

describe('VerificationCode', () => {
  test('renders correctly with default values', () => {
    shouldMatchEmotionSnapshot(<VerificationCode />)
  })

  test('renders correctly with initial value and placeholder and 6 fields', () => {
    shouldMatchEmotionSnapshot(
      <VerificationCode fields={6} initialValue="13" placeholder="0037" />,
    )
  })

  test('should handle paste with no overflowing values', () => {
    shouldMatchEmotionSnapshot(
      <VerificationCode type="number" fields={4} initialValue="1" />,
      {
        transform: ({ getByDisplayValue }) => {
          pasteEventWithValue(getByDisplayValue('1'), '1234')
        },
      },
    )
  })

  test('should handle and replace non number with "" when type is number', () => {
    shouldMatchEmotionSnapshot(
      <VerificationCode fields={4} initialValue="1" />,
      {
        transform: ({ getByDisplayValue }) => {
          pasteEventWithValue(getByDisplayValue('1'), '1a34')
        },
      },
    )
  })

  test('should handle paste with overflowing values', () => {
    shouldMatchEmotionSnapshot(
      <VerificationCode fields={4} initialValue="12" />,
      {
        transform: ({ getByDisplayValue }) => {
          pasteEventWithValue(getByDisplayValue('1'), '123456')
        },
      },
    )
  })

  test('should handle paste with overflowing values at different index than 0', () => {
    shouldMatchEmotionSnapshot(
      <VerificationCode fields={4} initialValue="12" />,
      {
        transform: ({ getByDisplayValue }) => {
          pasteEventWithValue(getByDisplayValue('2'), '123456')
        },
      },
    )
  })

  test('should handle paste when type is not number', () => {
    shouldMatchEmotionSnapshot(
      <VerificationCode type="text" fields={6} initialValue="12" />,
      {
        transform: ({ getByDisplayValue }) => {
          pasteEventWithValue(getByDisplayValue('2'), 'h23a*6')
        },
      },
    )
  })
})
