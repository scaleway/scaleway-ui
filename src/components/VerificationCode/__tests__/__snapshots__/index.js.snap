// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerificationCode renders correctly with default values 1`] = `
<DocumentFragment>
  .emotion-1 {
  border: solid 1px #dce1eb;
  font-size: 24px;
  color: #4a4f62;
  text-align: center;
  border-radius: 4px;
  margin-right: 8px;
  width: 56px;
  height: 64px;
}

.emotion-1:last-child {
  margin-right: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #dce1eb;
}

.emotion-1::-moz-placeholder {
  color: #dce1eb;
}

.emotion-1:-ms-input-placeholder {
  color: #dce1eb;
}

.emotion-1::placeholder {
  color: #dce1eb;
}

<div
    class="emotion-0"
  >
    <input
      class="emotion-1"
      data-id="0"
      id="verification-code-0"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value=""
    />
    <input
      class="emotion-1"
      data-id="1"
      id="verification-code-1"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value=""
    />
    <input
      class="emotion-1"
      data-id="2"
      id="verification-code-2"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value=""
    />
    <input
      class="emotion-1"
      data-id="3"
      id="verification-code-3"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`VerificationCode renders correctly with initial value and placeholder and 6 fields 1`] = `
<DocumentFragment>
  .emotion-1 {
  border: solid 1px #dce1eb;
  font-size: 24px;
  color: #4a4f62;
  text-align: center;
  border-radius: 4px;
  margin-right: 8px;
  width: 56px;
  height: 64px;
}

.emotion-1:last-child {
  margin-right: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #dce1eb;
}

.emotion-1::-moz-placeholder {
  color: #dce1eb;
}

.emotion-1:-ms-input-placeholder {
  color: #dce1eb;
}

.emotion-1::placeholder {
  color: #dce1eb;
}

<div
    class="emotion-0"
  >
    <input
      class="emotion-1"
      data-id="0"
      id="verification-code-0"
      pattern="[0-9]*"
      placeholder="0"
      type="tel"
      value="1"
    />
    <input
      class="emotion-1"
      data-id="1"
      id="verification-code-1"
      pattern="[0-9]*"
      placeholder="0"
      type="tel"
      value="3"
    />
    <input
      class="emotion-1"
      data-id="2"
      id="verification-code-2"
      pattern="[0-9]*"
      placeholder="3"
      type="tel"
      value=""
    />
    <input
      class="emotion-1"
      data-id="3"
      id="verification-code-3"
      pattern="[0-9]*"
      placeholder="7"
      type="tel"
      value=""
    />
    <input
      class="emotion-1"
      data-id="4"
      id="verification-code-4"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value=""
    />
    <input
      class="emotion-1"
      data-id="5"
      id="verification-code-5"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`VerificationCode should handle and replace non number with "" when type is number 1`] = `
<DocumentFragment>
  .emotion-1 {
  border: solid 1px #dce1eb;
  font-size: 24px;
  color: #4a4f62;
  text-align: center;
  border-radius: 4px;
  margin-right: 8px;
  width: 56px;
  height: 64px;
}

.emotion-1:last-child {
  margin-right: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #dce1eb;
}

.emotion-1::-moz-placeholder {
  color: #dce1eb;
}

.emotion-1:-ms-input-placeholder {
  color: #dce1eb;
}

.emotion-1::placeholder {
  color: #dce1eb;
}

<div
    class="emotion-0"
  >
    <input
      class="emotion-1"
      data-id="0"
      id="verification-code-0"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="1"
    />
    <input
      class="emotion-1"
      data-id="1"
      id="verification-code-1"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value=""
    />
    <input
      class="emotion-1"
      data-id="2"
      id="verification-code-2"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="3"
    />
    <input
      class="emotion-1"
      data-id="3"
      id="verification-code-3"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="4"
    />
  </div>
</DocumentFragment>
`;

exports[`VerificationCode should handle paste when type is not number 1`] = `
<DocumentFragment>
  .emotion-1 {
  border: solid 1px #dce1eb;
  font-size: 24px;
  color: #4a4f62;
  text-align: center;
  border-radius: 4px;
  margin-right: 8px;
  width: 56px;
  height: 64px;
}

.emotion-1:last-child {
  margin-right: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #dce1eb;
}

.emotion-1::-moz-placeholder {
  color: #dce1eb;
}

.emotion-1:-ms-input-placeholder {
  color: #dce1eb;
}

.emotion-1::placeholder {
  color: #dce1eb;
}

<div
    class="emotion-0"
  >
    <input
      class="emotion-1"
      data-id="0"
      id="verification-code-0"
      placeholder=""
      type="text"
      value="1"
    />
    <input
      class="emotion-1"
      data-id="1"
      id="verification-code-1"
      placeholder=""
      type="text"
      value="h"
    />
    <input
      class="emotion-1"
      data-id="2"
      id="verification-code-2"
      placeholder=""
      type="text"
      value="2"
    />
    <input
      class="emotion-1"
      data-id="3"
      id="verification-code-3"
      placeholder=""
      type="text"
      value="3"
    />
    <input
      class="emotion-1"
      data-id="4"
      id="verification-code-4"
      placeholder=""
      type="text"
      value="a"
    />
    <input
      class="emotion-1"
      data-id="5"
      id="verification-code-5"
      placeholder=""
      type="text"
      value="*"
    />
  </div>
</DocumentFragment>
`;

exports[`VerificationCode should handle paste with no overflowing values 1`] = `
<DocumentFragment>
  .emotion-1 {
  border: solid 1px #dce1eb;
  font-size: 24px;
  color: #4a4f62;
  text-align: center;
  border-radius: 4px;
  margin-right: 8px;
  width: 56px;
  height: 64px;
}

.emotion-1:last-child {
  margin-right: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #dce1eb;
}

.emotion-1::-moz-placeholder {
  color: #dce1eb;
}

.emotion-1:-ms-input-placeholder {
  color: #dce1eb;
}

.emotion-1::placeholder {
  color: #dce1eb;
}

<div
    class="emotion-0"
  >
    <input
      class="emotion-1"
      data-id="0"
      id="verification-code-0"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="1"
    />
    <input
      class="emotion-1"
      data-id="1"
      id="verification-code-1"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="2"
    />
    <input
      class="emotion-1"
      data-id="2"
      id="verification-code-2"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="3"
    />
    <input
      class="emotion-1"
      data-id="3"
      id="verification-code-3"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="4"
    />
  </div>
</DocumentFragment>
`;

exports[`VerificationCode should handle paste with overflowing values 1`] = `
<DocumentFragment>
  .emotion-1 {
  border: solid 1px #dce1eb;
  font-size: 24px;
  color: #4a4f62;
  text-align: center;
  border-radius: 4px;
  margin-right: 8px;
  width: 56px;
  height: 64px;
}

.emotion-1:last-child {
  margin-right: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #dce1eb;
}

.emotion-1::-moz-placeholder {
  color: #dce1eb;
}

.emotion-1:-ms-input-placeholder {
  color: #dce1eb;
}

.emotion-1::placeholder {
  color: #dce1eb;
}

<div
    class="emotion-0"
  >
    <input
      class="emotion-1"
      data-id="0"
      id="verification-code-0"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="1"
    />
    <input
      class="emotion-1"
      data-id="1"
      id="verification-code-1"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="2"
    />
    <input
      class="emotion-1"
      data-id="2"
      id="verification-code-2"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="3"
    />
    <input
      class="emotion-1"
      data-id="3"
      id="verification-code-3"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="4"
    />
  </div>
</DocumentFragment>
`;

exports[`VerificationCode should handle paste with overflowing values at different index than 0 1`] = `
<DocumentFragment>
  .emotion-1 {
  border: solid 1px #dce1eb;
  font-size: 24px;
  color: #4a4f62;
  text-align: center;
  border-radius: 4px;
  margin-right: 8px;
  width: 56px;
  height: 64px;
}

.emotion-1:last-child {
  margin-right: 0;
}

.emotion-1::-webkit-input-placeholder {
  color: #dce1eb;
}

.emotion-1::-moz-placeholder {
  color: #dce1eb;
}

.emotion-1:-ms-input-placeholder {
  color: #dce1eb;
}

.emotion-1::placeholder {
  color: #dce1eb;
}

<div
    class="emotion-0"
  >
    <input
      class="emotion-1"
      data-id="0"
      id="verification-code-0"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="1"
    />
    <input
      class="emotion-1"
      data-id="1"
      id="verification-code-1"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="1"
    />
    <input
      class="emotion-1"
      data-id="2"
      id="verification-code-2"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="2"
    />
    <input
      class="emotion-1"
      data-id="3"
      id="verification-code-3"
      pattern="[0-9]*"
      placeholder=""
      type="tel"
      value="3"
    />
  </div>
</DocumentFragment>
`;
